Userland/Libraries/LibThreading/BackgroundAction.h

```diff
@@ -34,55 +35,82 @@ class BackgroundActionBase {
 };
 
 template<typename Result>
-class BackgroundAction final : public Core::Object
-    , private BackgroundActionBase {
+class BackgroundAction final : private BackgroundActionBase
+    , public Core::Object {
     C_OBJECT(BackgroundAction);
 
 public:
-    void cancel()
-    {
-        m_cancelled = true;
-    }
-
-    bool is_cancelled() const
-    {
-        return m_cancelled;
-    }
+    // Promise is an implementation detail of BackgroundAction in order to communicate with EventLoop.
+    // All of the promise's callbacks and state are either managed by us or by EventLoop.
+    using Promise = Core::Promise<NonnullRefPtr<Core::Object>>;
 
     virtual ~BackgroundAction() = default;
 
+    Optional<Result> const& result() const { return m_result; }
+    Optional<Result>& result() { return m_result; }
+
+    void cancel() { m_cancelled = true; }
+    // If your action is long-running, you should periodically check the cancel state and possibly return early.
+    bool is_cancelled() const { return m_cancelled; }
```

Comment:
Not sure what our preference is in other places in the codebase, but `cancelled` is the UK spelling, while `canceled` is the US one.